fe/fe-core/src/main/java/com/starrocks/load/pipe/Pipe.java

```diff
@@ -0,0 +1,501 @@
+            // EOS
+            if (fileSource.eos()) {
+                changeState(State.FINISHED, true);
+                LOG.info("pipe {} finish all tasks, change state to {}", this, state);
+            }
+            return;
+        }
+
+        try {
+            lock.writeLock().lock();
+
+            // TODO: structural replace ?
+            String sqlTask = originSql.replaceAll("(?i)TABLE\\(.*\\)", buildFileSelectSource(piece));
+            long taskId = GlobalStateMgr.getCurrentState().getNextId();
+            PipeId pipeId = getPipeId();
+            String uniqueName = PipeTaskDesc.genUniqueTaskName(getName(), taskId, 0);
+            String dbName = GlobalStateMgr.getCurrentState().mayGetDb(pipeId.getDbId())
+                    .map(Database::getOriginName)
+                    .orElseThrow(() -> ErrorReport.buildSemanticException(ErrorCode.ERR_BAD_DB_ERROR));
+            PipeTaskDesc taskDesc = new PipeTaskDesc(taskId, uniqueName, dbName, sqlTask, piece);
+            taskDesc.setErrorLimit(FAILED_TASK_THRESHOLD);
+
+            runningTasks.put(taskId, taskDesc);
+            LOG.debug("pipe {} build task: {}", name, taskDesc);
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    /**
+     * Schedule runnable tasks
+     * // TODO: async execution
+     */
+    private void scheduleRunnableTasks() {
+        if (MapUtils.isEmpty(runningTasks)) {
+            return;
+        }
+
+        for (PipeTaskDesc task : runningTasks.values()) {
+            executeTask(task);
+        }
+    }
+
+    /**
+     * Clean up FINISHED/FAILED tasks
+     */
+    private void finalizeTasks() {
+        List<Long> removeTaskId = new ArrayList<>();
+        try {
+            lock.writeLock().lock();
+
+            for (PipeTaskDesc task : runningTasks.values()) {
+                if (task.isFinished() || task.tooManyErrors()) {
+                    removeTaskId.add(task.getId());
+                    FilePipePiece piece = task.getPiece();
+                    ((FilePipeSource) pipeSource).finishPiece(piece, task.getState());
+                }
+                if (task.isError()) {
+                    failedTaskExecutionCount++;
+                    if (failedTaskExecutionCount > FAILED_TASK_THRESHOLD) {
+                        changeState(State.ERROR, false);
+                    }
+                }
+                if (task.isFinished()) {
+                    FilePipePiece piece = task.getPiece();
+                    loadStatus.loadFiles++;
+                    loadStatus.loadBytes += piece.getFile().getSize();
+                    // TODO: fill data rows
+                    loadStatus.loadRows += 1;
+                }
+            }
+            for (long taskId : removeTaskId) {
+                runningTasks.remove(taskId);
+            }
+        } finally {
+            lock.writeLock().unlock();
+        }
+
+        if (CollectionUtils.isNotEmpty(removeTaskId)) {
+            persistPipe();
```

Comment:
Is it necessary to guarantee the atomicity for updating the file list in `FilePipeSource#finishPiece` and persisting the pipe?

Reply:
good point. 

In theory we need to keep the consistency of `Pipe::LoadStatus` and `PipeSource::FileList`, specially the data rows of finished files should be exact.

Current implementation cannot guarantee it, which should be considered when implement FileList persistence and task recovery.

Reply:
The atomicity of `FileList` update and `Pipe::LoadStatus` should be guaranteed